core/net/rpl/rpl-nbr-policy.c

```diff
@@ -0,0 +1,252 @@
+    linkaddr_t *lladdr = nbr_table_get_lladdr(ds6_neighbors, nbr);
+    is_used = 0;
+
+    parent = rpl_get_parent((uip_lladdr_t *)lladdr);
+    if(parent != NULL) {
+      num_parents++;
+      is_used++;
+
+      if(parent->dag != NULL && parent->dag->preferred_parent == parent) {
+        /* This is the preferred parent for the DAG and must not be removed */
+
+        /* Note: this assumes that only RPL adds default routes. */
+
+      } else if(worst_rank < INFINITE_RANK &&
+                parent->rank > 0 &&
+                parent->dag != NULL &&
+                parent->dag->instance != NULL &&
+                (rank = parent->dag->instance->of->calculate_rank(parent, 0)) > worst_rank) {
+        /* This is the worst-rank neighbor - this is a good candidate for removal */
+        if(uip_ds6_route_is_nexthop((uip_lladdr_t *)lladdr) == 0) {
+          worst_rank = rank;
+          worst_rank_nbr = lladdr;
+        } else {
+          printf("*** Can not use this as worst rank as it is a next hop\n");
+        }
+      }
+    }
+
+    /* Check if this neighbor is used as nexthop and therefor being a
+       RPL child. */
+    if(uip_ds6_route_is_nexthop((uip_lladdr_t *)lladdr) != 0) {
+      is_used++;
+      num_children++;
+    }
+
+    if(is_used == 0) {
+      /* This neighbor is neither parent or child and can be safely removed */
+      worst_rank_nbr = lladdr;
+      worst_rank = INFINITE_RANK;
+    } else if(is_used > 1) {
+      PRINTF("NBR-POLICY: *** Neighbor is both child and candidate parent: ");
+      PRINTLLADDR((uip_lladdr_t *)lladdr);
+      PRINTF("\n");
+    }
+
+    nbr = nbr_table_next(ds6_neighbors, nbr);
+    num_used++;
+  }
+  /* how many more IP neighbors can be have? */
+  num_free = NBR_TABLE_MAX_NEIGHBORS - num_used;
+
+  PRINTF("NBR-POLICY: Free: %d, Children: %d, Parents: %d Routes: %d\n",
+	 num_free, num_children, num_parents, uip_ds6_route_num_routes());
+}
+/*---------------------------------------------------------------------------*/
+/* Called whenever we get a unicast DIS - e.g. someone that already
+   have this node in its table - since it is a unicast */
+const linkaddr_t *
+find_removable_dis(uip_ipaddr_t *from)
+{
+
+  update_nbr();
+  if(num_free > 0) {
+    printf("num-free > 0 = %d", num_free);
+    printf("**** Should remove unused elements but can not... \n");
+    /*    return 1; */
+  }
+  if(num_children < MAX_CHILDREN) {
+    return worst_rank_nbr;
+  }
+  return NULL;
+}
+/*---------------------------------------------------------------------------*/
+const linkaddr_t *
+find_removable_dio(uip_ipaddr_t *from, rpl_dio_t *dio)
+{
+  rpl_instance_t *instance;
+  rpl_rank_t rank;
+
+  update_nbr();
+
+  instance = rpl_get_instance(dio->instance_id);
+  if(instance == NULL || instance->current_dag == NULL) {
+    PRINTF("Did not find instance id: %d\n", dio->instance_id);
+    return NULL;
+  }
+
+  /* Add the new neighbor only if it is better than the preferred parent. */
```

Comment:
Yes, you are right - it is comparing with the currently worst neighbor / potential parent.